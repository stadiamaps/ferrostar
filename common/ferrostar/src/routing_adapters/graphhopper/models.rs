use serde::{Deserialize};
use alloc::collections::BTreeMap as HashMap;
use crate::models::{ GeographicCoordinate };

#[derive(Deserialize, Debug)]
pub struct GraphHopperRouteResponse {
    pub paths: Vec<GraphHopperPath>,
    pub message: Option<String>
}

#[derive(Deserialize, Debug)]
#[serde(untagged)]
pub enum DetailEntryValue {
    Int(i32),
    Str(String),
    Bool(bool),
    Float(f64),
}

#[derive(Deserialize, Debug)]
#[serde(from = "Vec<serde_json::Value>")]
pub struct DetailEntry {
    pub start_index: usize,
    pub end_index: usize,
    pub value: Option<DetailEntryValue>,
}

impl From<Vec<serde_json::Value>> for DetailEntry {
    fn from(vec: Vec<serde_json::Value>) -> Self {
        let start_index = vec[0].as_i64().unwrap_or_default() as usize;
        let end_index = vec[1].as_i64().unwrap_or_default() as usize;

        let value = match &vec[2] {
            serde_json::Value::String(s) => Some(DetailEntryValue::Str(s.clone())),
            serde_json::Value::Number(n) => {
                if n.is_i64() {
                    Some(DetailEntryValue::Int(n.as_i64().unwrap() as i32))
                } else {
                    Some(DetailEntryValue::Float(n.as_f64().unwrap()))
                }
            },
            serde_json::Value::Bool(b) => Some(DetailEntryValue::Bool(*b)),
            serde_json::Value::Null => None,
            _ => None,
        };

        DetailEntry { start_index, end_index, value }
    }
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperPath {
    pub distance: f64,
    pub time: f64,
    pub bbox: Vec<f64>,
    pub instructions: Vec<GraphHopperInstruction>,
    pub details: HashMap<String, Vec<DetailEntry>>,
    pub points: String, // encoded polyline
    pub points_encoded:	bool,
    pub points_encoded_multiplier: f64,
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperInstruction {
    pub distance: f64,
    pub time: f64,
    pub heading: Option<f64>,
    pub exit_number: Option<u32>,
    pub turn_angle: Option<f64>,
    pub sign: i32,
    pub interval: Vec<usize>,
    pub text: String,
    pub street_ref: Option<String>,
    pub street_name: String,
}

// temporary structure until we understand internal maxspeed handling
pub struct MaxSpeedEntry {
    pub geometry: Vec<GeographicCoordinate>,
    pub speed_limit: Option<f64>,
    pub unit: String,
}
