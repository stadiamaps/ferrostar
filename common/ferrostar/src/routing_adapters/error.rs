#[cfg(feature = "alloc")]
use alloc::string::{String, ToString};

// TODO: This implementation seems less than ideal. In particular, it hides what sort of JSON error occurred due to an apparent bug in UniFFI.
// The trouble appears to be with generating "flat" enum bindings that are used with callback
// interfaces when the underlying actually has fields.
#[derive(Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
pub enum InstantiationError {
    #[cfg_attr(
        feature = "std",
        error("Error parsing the JSON options for the request.")
    )]
    OptionsJsonParseError,
}

// TODO: See comment above
#[derive(Debug)]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
pub enum RoutingRequestGenerationError {
    #[cfg_attr(
        feature = "std",
        error("Too few waypoints were provided to compute a route.")
    )]
    NotEnoughWaypoints,
    #[cfg_attr(feature = "std", error("Error generating JSON for the request."))]
    JsonError,
    #[cfg_attr(
        feature = "std",
        error("An unknown error generating a request was raised in foreign code.")
    )]
    UnknownRequestGenerationError,
}

#[cfg(feature = "uniffi")]
impl From<uniffi::UnexpectedUniFFICallbackError> for RoutingRequestGenerationError {
    fn from(_: uniffi::UnexpectedUniFFICallbackError) -> RoutingRequestGenerationError {
        RoutingRequestGenerationError::UnknownRequestGenerationError
    }
}

impl From<serde_json::Error> for InstantiationError {
    fn from(_: serde_json::Error) -> Self {
        InstantiationError::OptionsJsonParseError
    }
}

impl From<serde_json::Error> for RoutingRequestGenerationError {
    fn from(_: serde_json::Error) -> Self {
        RoutingRequestGenerationError::JsonError
    }
}

#[derive(Debug)]
#[cfg_attr(feature = "std", derive(thiserror::Error))]
#[cfg_attr(feature = "uniffi", derive(uniffi::Error))]
pub enum ParsingError {
    // TODO: Unable to find route and other common errors
    #[cfg_attr(feature = "std", error("Failed to parse route json object: {error}."))]
    InvalidRouteObject { error: String },
    #[cfg_attr(feature = "std", error("Failed to parse route geometry: {error}."))]
    InvalidGeometry { error: String },
    #[cfg_attr(feature = "std", error("Failed to parse annotations: {error}."))]
    MalformedAnnotations { error: String },
    #[cfg_attr(
        feature = "std",
        error("Routing adapter returned an unexpected status code: {code}.")
    )]
    InvalidStatusCode { code: String },
    #[cfg_attr(
        feature = "std",
        error("An unknown error parsing a response was raised in foreign code.")
    )]
    UnknownParsingError,
}

#[cfg(feature = "uniffi")]
impl From<uniffi::UnexpectedUniFFICallbackError> for ParsingError {
    fn from(_: uniffi::UnexpectedUniFFICallbackError) -> ParsingError {
        ParsingError::UnknownParsingError
    }
}

impl From<serde_json::Error> for ParsingError {
    fn from(e: serde_json::Error) -> Self {
        ParsingError::InvalidRouteObject {
            error: e.to_string(),
        }
    }
}
