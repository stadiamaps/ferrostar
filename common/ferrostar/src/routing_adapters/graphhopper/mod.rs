use crate::models::{UserLocation, Waypoint};

use crate::routing_adapters::error::{InstantiationError, RoutingRequestGenerationError};
use crate::routing_adapters::{RouteRequest, RouteRequestGenerator};

use serde_json::{json, Map, Value as JsonValue};
#[cfg(feature = "std")]
use std::collections::HashMap;

#[derive(Debug)]
pub struct GraphHopperHttpRequestGenerator {
    endpoint_url: String,
    profile: String,
    options: Map<String, JsonValue>,
}

impl GraphHopperHttpRequestGenerator {
    pub fn new(endpoint_url: String, profile: String, options: Map<String, JsonValue>) -> Self {
        Self {
            endpoint_url,
            profile,
            options,
        }
    }

    pub fn with_options_json(
        endpoint_url: String,
        profile: String,
        options_json: Option<&str>,
    ) -> Result<Self, InstantiationError> {
        let parsed_options = match options_json {
            // TODO: Another error variant
            Some(options) => serde_json::from_str::<JsonValue>(options)?
                .as_object()
                .ok_or(InstantiationError::OptionsJsonParseError)?
                .to_owned(),
            None => Map::new(),
        };
        Ok(Self {
            endpoint_url,
            profile,
            options: parsed_options,
        })
    }
}

impl RouteRequestGenerator for GraphHopperHttpRequestGenerator {
    fn generate_request(
        &self,
        user_location: UserLocation,
        waypoints: Vec<Waypoint>,
    ) -> Result<RouteRequest, RoutingRequestGenerationError> {
        if waypoints.is_empty() {
            Err(RoutingRequestGenerationError::NotEnoughWaypoints)
        } else {
            let headers =
                HashMap::from([("Content-Type".to_string(), "application/json".to_string())]);

            let mut points: Vec<Vec<f64>> = vec![vec![
                user_location.coordinates.lng,
                user_location.coordinates.lat,
            ]];

            points.extend(
                waypoints
                    .iter()
                    .map(|waypoint| vec![waypoint.coordinate.lng, waypoint.coordinate.lat]),
            );

            let mut args = json!({
                "profile": &self.profile,
                "points": points,
                "locale": "en",
                "type": "mapbox",
                "voice_units": "metric",
            });

            for (k, v) in &self.options {
                args[k] = v.clone();
            }

            let body = serde_json::to_vec(&args)?;
            Ok(RouteRequest::HttpPost {
                url: self.endpoint_url.clone(),
                headers,
                body,
            })
        }
    }
}
