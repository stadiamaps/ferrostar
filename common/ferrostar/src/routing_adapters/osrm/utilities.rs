use super::models::AnyAnnotation;
use crate::models::AnyAnnotationValue;
use crate::routing_adapters::error::ParsingError;
use serde_json::Value;
use std::collections::HashMap;

/// Gets a slice of the route's annotations array.
///
/// Returns a [`ParsingError`] if the annotations are not present or the slice is out of bounds.
pub(crate) fn get_annotation_slice(
    annotations: Option<Vec<AnyAnnotationValue>>,
    start_index: usize,
    end_index: usize,
) -> Result<Vec<AnyAnnotationValue>, ParsingError> {
    annotations
        .ok_or(ParsingError::MalformedAnnotations {
            error: "No annotations".to_string(),
        })?
        .get(start_index..end_index)
        .ok_or(ParsingError::MalformedAnnotations {
            error: "Annotations slice index out of bounds ({start_index}..{end_index})".to_string(),
        })
        .map(<[AnyAnnotationValue]>::to_vec)
}

/// Converts the the OSRM-style annotation object consisting of separate arrays
/// to a single vector of parsed objects (one for each coordinate pair).
pub(crate) fn zip_annotations(annotation: AnyAnnotation) -> Vec<AnyAnnotationValue> {
    let source: HashMap<String, Vec<Value>> = annotation.values;

    // Get the length of the array (assumed to be the same for all annotations)
    let length = source.values().next().map_or(0, Vec::len);

    return (0..length)
        .map(|i| {
            source
                .iter()
                .filter_map(|(key, values)| {
                    // Values is the vector at a specific key in the original annotations object.
                    values
                        .get(i) // For each key, get the value at the index i.
                        .map(|v| (key.clone(), v.clone())) // Convert the key and value to a tuple.
                })
                .collect::<HashMap<String, Value>>() // Collect the key-value pairs into a hashmap.
        })
        .map(|value| AnyAnnotationValue { value })
        .collect::<Vec<AnyAnnotationValue>>();
}

#[cfg(test)]
mod test {

    use serde_json::Map;

    use super::*;

    #[test]
    fn test_zip_annotation() {
        let json_str = r#"{
            "distance": [1.2, 2.24, 3.24],
            "duration": [4, 5, 6],
            "speed": [10, 11, 12],
            "max_speed": [{
              "speed": 56,
              "unit": "km/h"
            }, {
              "speed": 12,
              "unit": "mi/h"
            }, {
              "unknown": true
            }],
            "traffic": ["bad", "ok", "good"],
            "construction": [null, true, null]
        }"#;

        let json_value: Map<String, Value> = serde_json::from_str(json_str).unwrap();
        let values: HashMap<String, Vec<Value>> = json_value
            .iter()
            .map(|(k, v)| (k.to_string(), v.as_array().unwrap().clone()))
            .collect();

        // Construct the annotation object.
        let annotation = AnyAnnotation { values };
        let result = zip_annotations(annotation);

        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(result);
        });
    }
}
