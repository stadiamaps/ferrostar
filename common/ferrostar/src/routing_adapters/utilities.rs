use polyline::decode_polyline;

use crate::models::GeographicCoordinate;

use super::error::ParsingError;

/// Parse a polyline-encoded geometry string into a list of geographic coordinates.
/// If the polyline cannot be decoded, a [`ParsingError`] results.
pub fn get_coordinates_from_geometry(
    geometry: &str,
    polyline_precision: u32,
) -> Result<Vec<GeographicCoordinate>, ParsingError> {
    let linestring = decode_polyline(geometry, polyline_precision).map_err(|error| {
        ParsingError::InvalidGeometry {
            error: error.to_string(),
        }
    })?;

    // TODO: Trait for this common pattern?
    let linestring = linestring
        .coords()
        .map(|coord| GeographicCoordinate::from(*coord))
        .collect();

    Ok(linestring)
}
