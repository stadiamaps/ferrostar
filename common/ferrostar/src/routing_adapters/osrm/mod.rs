//! Response parsing for OSRM-compatible JSON (including Stadia Maps, Valhalla, Mapbox, etc.).

pub(crate) mod models;
pub mod utilities;

use super::RouteResponseParser;
use crate::models::{
    AnyAnnotationValue, GeographicCoordinate, Incident, LaneInfo, RouteStep, SpokenInstruction,
    VisualInstruction, VisualInstructionContent, Waypoint, WaypointKind,
};
use crate::routing_adapters::utilities::get_coordinates_from_geometry;
use crate::routing_adapters::{
    osrm::models::{
        Route as OsrmRoute, RouteResponse, RouteStep as OsrmRouteStep, Waypoint as OsrmWaypoint,
    },
    ParsingError, Route,
};
#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::{string::ToString, vec, vec::Vec};
use geo::BoundingRect;
use models::BannerContent;
use polyline::decode_polyline;
use utilities::get_annotation_slice;
use uuid::Uuid;

/// A response parser for OSRM-compatible routing backends.
///
/// The parser is NOT limited to only the standard OSRM format; many Valhalla/Mapbox tags are also
/// parsed and are included in the final route.
#[derive(Debug)]
pub struct OsrmResponseParser {
    polyline_precision: u32,
}

impl OsrmResponseParser {
    pub fn new(polyline_precision: u32) -> Self {
        Self { polyline_precision }
    }
}

impl RouteResponseParser for OsrmResponseParser {
    fn parse_response(&self, response: Vec<u8>) -> Result<Vec<Route>, ParsingError> {
        let res: RouteResponse = serde_json::from_slice(&response)?;

        if res.code == "Ok" {
            res.routes
                .iter()
                .map(|route| Route::from_osrm(route, &res.waypoints, self.polyline_precision))
                .collect::<Result<Vec<_>, _>>()
        } else {
            Err(ParsingError::InvalidStatusCode { code: res.code })
        }
    }
}

impl Route {
    pub fn from_osrm(
        route: &OsrmRoute,
        waypoints: &[OsrmWaypoint],
        polyline_precision: u32,
    ) -> Result<Self, ParsingError> {
        let via_waypoint_indices: Vec<_> = route
            .legs
            .iter()
            .flat_map(|leg| leg.via_waypoints.iter().map(|via| via.waypoint_index))
            .collect();

        let waypoints: Vec<_> = waypoints
            .iter()
            .enumerate()
            .map(|(idx, waypoint)| Waypoint {
                coordinate: GeographicCoordinate {
                    lat: waypoint.location.latitude(),
                    lng: waypoint.location.longitude(),
                },
                kind: if via_waypoint_indices.contains(&idx) {
                    WaypointKind::Via
                } else {
                    WaypointKind::Break
                },
            })
            .collect();

        let linestring = decode_polyline(&route.geometry, polyline_precision).map_err(|error| {
            ParsingError::InvalidGeometry {
                error: error.to_string(),
            }
        })?;
        if let Some(bbox) = linestring.bounding_rect() {
            let geometry: Vec<GeographicCoordinate> = linestring
                .coords()
                .map(|coord| GeographicCoordinate::from(*coord))
                .collect();

            let steps = route
                .legs
                .iter()
                .flat_map(|leg| {
                    // Converts all single value annotation vectors into a single vector witih a value object.
                    let annotations = leg
                        .annotation
                        .as_ref()
                        .map(|leg_annotation| utilities::zip_annotations(leg_annotation.clone()));

                    // Convert all incidents into a vector of Incident objects.
                    let incident_items = leg
                        .incidents
                        .iter()
                        .map(Incident::from)
                        .collect::<Vec<Incident>>();

                    // Index for the annotations slice
                    let mut start_index: usize = 0;

                    leg.steps.iter().map(move |step| {
                        let step_geometry =
                            get_coordinates_from_geometry(&step.geometry, polyline_precision)?;

                        // Slice the annotations for the current step.
                        // The annotations array represents segments between coordinates.
                        //
                        // 1. Annotations should never repeat.
                        // 2. Each step has one less annotation than coordinate.
                        // 3. The last step never has annotations as it's two of the route's last coordinate (duplicate).
                        let step_index_len = step_geometry.len() - 1_usize;
                        let end_index = start_index + step_index_len;

                        let annotation_slice =
                            get_annotation_slice(annotations.clone(), start_index, end_index).ok();

                        let relevant_incidents_slice = incident_items
                            .iter()
                            .filter(|incident| {
                                let incident_start = incident.geometry_index_start as usize;

                                match incident.geometry_index_end {
                                    Some(end) => {
                                        let incident_end = end as usize;
                                        incident_start >= start_index && incident_end <= end_index
                                    }
                                    None => {
                                        incident_start >= start_index && incident_start <= end_index
                                    }
                                }
                            })
                            .map(|incident| {
                                let mut adjusted_incident = incident.clone();
                                if adjusted_incident.geometry_index_start - start_index as u64 > 0 {
                                    adjusted_incident.geometry_index_start -= start_index as u64;
                                } else {
                                    adjusted_incident.geometry_index_start = 0;
                                }

                                if let Some(end) = adjusted_incident.geometry_index_end {
                                    let adjusted_end = end - start_index as u64;
                                    adjusted_incident.geometry_index_end =
                                        Some(if adjusted_end > end_index as u64 {
                                            end_index as u64
                                        } else {
                                            adjusted_end
                                        });
                                }
                                adjusted_incident
                            })
                            .collect::<Vec<Incident>>();

                        start_index = end_index;

                        RouteStep::from_osrm_and_geom(
                            step,
                            step_geometry,
                            annotation_slice,
                            relevant_incidents_slice,
                        )
                    })
                })
                .collect::<Result<Vec<_>, _>>()?;

            Ok(Route {
                geometry,
                bbox: bbox.into(),
                distance: route.distance,
                waypoints: waypoints.clone(),
                steps,
            })
        } else {
            Err(ParsingError::InvalidGeometry {
                error: "Bounding box could not be calculated".to_string(),
            })
        }
    }
}

impl RouteStep {
    fn extract_exit_numbers(banner_content: &BannerContent) -> Vec<String> {
        banner_content
            .components
            .iter()
            .filter(|component| component.component_type.as_deref() == Some("exit-number"))
            .filter_map(|component| component.text.clone())
            .collect()
    }

    fn from_osrm_and_geom(
        value: &OsrmRouteStep,
        geometry: Vec<GeographicCoordinate>,
        annotations: Option<Vec<AnyAnnotationValue>>,
        incidents: Vec<Incident>,
    ) -> Result<Self, ParsingError> {
        let visual_instructions = value
            .banner_instructions
            .iter()
            .map(|banner| VisualInstruction {
                primary_content: VisualInstructionContent {
                    text: banner.primary.text.clone(),
                    maneuver_type: banner.primary.maneuver_type,
                    maneuver_modifier: banner.primary.maneuver_modifier,
                    roundabout_exit_degrees: banner.primary.roundabout_exit_degrees,
                    lane_info: None,
                    exit_numbers: Self::extract_exit_numbers(&banner.primary),
                },
                secondary_content: banner.secondary.as_ref().map(|secondary| {
                    VisualInstructionContent {
                        text: secondary.text.clone(),
                        maneuver_type: secondary.maneuver_type,
                        maneuver_modifier: secondary.maneuver_modifier,
                        roundabout_exit_degrees: banner.primary.roundabout_exit_degrees,
                        lane_info: None,
                        exit_numbers: Self::extract_exit_numbers(&secondary),
                    }
                }),
                sub_content: banner.sub.as_ref().map(|sub| VisualInstructionContent {
                    text: sub.text.clone(),
                    maneuver_type: sub.maneuver_type,
                    maneuver_modifier: sub.maneuver_modifier,
                    roundabout_exit_degrees: sub.roundabout_exit_degrees,
                    lane_info: {
                        let lane_infos: Vec<LaneInfo> = sub
                            .components
                            .iter()
                            .filter(|component| component.component_type.as_deref() == Some("lane"))
                            .map(|component| LaneInfo {
                                active: component.active.unwrap_or(false),
                                directions: component.directions.clone().unwrap_or_default(),
                                active_direction: component.active_direction.clone(),
                            })
                            .collect();

                        if lane_infos.is_empty() {
                            None
                        } else {
                            Some(lane_infos)
                        }
                    },
                    exit_numbers: Self::extract_exit_numbers(&sub),
                }),
                trigger_distance_before_maneuver: banner.distance_along_geometry,
            })
            .collect();

        let spoken_instructions = value
            .voice_instructions
            .iter()
            .map(|instruction| SpokenInstruction {
                text: instruction.announcement.clone(),
                ssml: instruction.ssml_announcement.clone(),
                trigger_distance_before_maneuver: instruction.distance_along_geometry,
                utterance_id: Uuid::new_v4(),
            })
            .collect();

        // Convert the annotations to a vector of json strings.
        // This allows us to safely pass the RouteStep through the FFI boundary.
        // The host platform can then parse an arbitrary annotation object.
        let annotations_as_strings: Option<Vec<String>> = annotations.map(|annotations_vec| {
            annotations_vec
                .iter()
                .map(|annotation| serde_json::to_string(annotation).unwrap())
                .collect()
        });

        let exits = match value.exits.clone() {
            Some(exit_text) => exit_text.split(';').map(|s| s.trim().to_string()).collect(),
            None => Vec::new(),
        };

        Ok(RouteStep {
            geometry,
            // TODO: Investigate using the haversine distance or geodesics to normalize.
            // Valhalla in particular is a bit nonstandard. See https://github.com/valhalla/valhalla/issues/1717
            distance: value.distance,
            duration: value.duration,
            road_name: value.name.clone(),
            exits,
            instruction: value.maneuver.get_instruction(),
            visual_instructions,
            spoken_instructions,
            annotations: annotations_as_strings,
            incidents,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const STANDARD_OSRM_POLYLINE6_RESPONSE: &str =
        include_str!("fixtures/standard_osrm_polyline6_response.json");
    const VALHALLA_OSRM_RESPONSE: &str = include_str!("fixtures/valhalla_osrm_response.json");
    const VALHALLA_OSRM_RESPONSE_VIA_WAYS: &str =
        include_str!("fixtures/valhalla_osrm_response_via_ways.json");
    const VALHALLA_EXTENDED_OSRM_RESPONSE: &str =
        include_str!("fixtures/valhalla_extended_osrm_response.json");
    const VALHALLA_OSRM_RESPONSE_WITH_EXITS: &str =
        include_str!("fixtures/valhalla_osrm_response_with_exit_info.json");

    #[test]
    fn parse_standard_osrm() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(STANDARD_OSRM_POLYLINE6_RESPONSE.into())
            .expect("Unable to parse OSRM response");
        insta::assert_yaml_snapshot!(routes);
    }

    #[test]
    fn parse_valhalla_osrm() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }

    #[test]
    fn parse_valhalla_osrm_with_via_ways() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE_VIA_WAYS.into())
            .expect("Unable to parse Valhalla OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }

    #[test]
    fn parse_valhalla_asserting_annotation_lengths() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla OSRM response");

        // Loop through every step and validate that the length of the annotations
        // matches the length of the geometry minus one. This is because each annotation
        // represents a segment between two coordinates.
        for (route_index, route) in routes.iter().enumerate() {
            for (step_index, step) in route.steps.iter().enumerate() {
                if step_index == route.steps.len() - 1 {
                    // The arrival step is 2 of the same coordinates.
                    // So annotations will be None.
                    assert_eq!(step.annotations, None);
                    continue;
                }

                let step = step.clone();
                let annotations = step.annotations.expect("No annotations");
                assert_eq!(
                    annotations.len(),
                    step.geometry.len() - 1,
                    "Route {}, Step {}",
                    route_index,
                    step_index
                );
            }
        }
    }

    #[test]
    fn parse_valhalla_asserting_sub_maneuvers() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_EXTENDED_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla Extended OSRM response");

        // Collect all sub_contents into a vector
        let sub_contents: Vec<_> = routes
            .iter()
            .flat_map(|route| &route.steps)
            .filter_map(|step| {
                step.visual_instructions
                    .iter()
                    .find_map(|instruction: &VisualInstruction| instruction.sub_content.as_ref())
            })
            .collect();

        // Assert that there's exactly one sub maneuver instructions as is the case in the test data
        assert_eq!(
            sub_contents.len(),
            1,
            "Expected exactly one sub banner instructions"
        );

        if let Some(sub_content) = sub_contents.first() {
            // Ensure that there are 4 pieces of lane information in the sub banner instructions
            if let Some(lane_info) = &sub_content.lane_info {
                assert_eq!(lane_info.len(), 4);
            } else {
                panic!("Expected lane information, but could not find it");
            }
        } else {
            panic!("No sub banner instructions found in any of the steps")
        }
    }

    #[test]
    fn parse_osrm_with_exits() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE_WITH_EXITS.into())
            .expect("Unable to parse OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }
}
