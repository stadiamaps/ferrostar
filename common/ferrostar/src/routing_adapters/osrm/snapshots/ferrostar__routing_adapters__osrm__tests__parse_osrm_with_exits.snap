---
source: ferrostar/src/routing_adapters/osrm/mod.rs
expression: routes
---
- geometry:
    - lat: 45.377106
      lng: -122.584167
    - lat: 45.376869
      lng: -122.584504
    - lat: 45.376527
      lng: -122.584983
    - lat: 45.375948
      lng: -122.585794
    - lat: 45.375411
      lng: -122.586549
    - lat: 45.374846
      lng: -122.587677
    - lat: 45.374808
      lng: -122.587796
    - lat: 45.374781
      lng: -122.587928
    - lat: 45.374774
      lng: -122.588028
    - lat: 45.37478
      lng: -122.588136
    - lat: 45.374789
      lng: -122.588209
    - lat: 45.374812
      lng: -122.588294
    - lat: 45.374849
      lng: -122.588384
    - lat: 45.37488
      lng: -122.588451
    - lat: 45.374919
      lng: -122.588512
    - lat: 45.374978
      lng: -122.588569
    - lat: 45.375024
      lng: -122.58861
    - lat: 45.375075
      lng: -122.58864
    - lat: 45.375143
      lng: -122.588661
    - lat: 45.375324
      lng: -122.588722
    - lat: 45.375415
      lng: -122.588676
    - lat: 45.375483
      lng: -122.588617
    - lat: 45.375559
      lng: -122.58852
    - lat: 45.37561
      lng: -122.588432
    - lat: 45.375657
      lng: -122.58831
    - lat: 45.375703
      lng: -122.588114
    - lat: 45.375631
      lng: -122.58789
    - lat: 45.375576
      lng: -122.587713
    - lat: 45.375506
      lng: -122.58758
    - lat: 45.375468
      lng: -122.587524
    - lat: 45.37531
      lng: -122.587349
    - lat: 45.375003
      lng: -122.587226
    - lat: 45.374495
      lng: -122.587017
    - lat: 45.374137
      lng: -122.586867
    - lat: 45.373646
      lng: -122.586686
    - lat: 45.373572
      lng: -122.586658
    - lat: 45.373194
      lng: -122.586506
    - lat: 45.372855
      lng: -122.586367
    - lat: 45.372463
      lng: -122.586234
    - lat: 45.372156
      lng: -122.586125
    - lat: 45.371846
      lng: -122.586016
    - lat: 45.371508
      lng: -122.585901
  bbox:
    sw:
      lat: 45.371508
      lng: -122.588722
    ne:
      lat: 45.377106
      lng: -122.584167
  distance: 1072.913
  waypoints:
    - coordinate:
        lat: 45.377106
        lng: -122.584167
      kind: Break
    - coordinate:
        lat: 45.371508
        lng: -122.585901
      kind: Break
  steps:
    - geometry:
        - lat: 45.377106
          lng: -122.584167
        - lat: 45.376869
          lng: -122.584504
        - lat: 45.376527
          lng: -122.584983
        - lat: 45.375948
          lng: -122.585794
        - lat: 45.375411
          lng: -122.586549
      distance: 264.647
      duration: 10.705
      road_name: ""
      exits: []
      instruction: Drive southwest on I 205 South/OR 213 South.
      visual_instructions:
        - primary_content:
            text: "OR 213 South: Oregon City, Molalla"
            maneuver_type: off ramp
            maneuver_modifier: slight right
            roundabout_exit_degrees: ~
            lane_info: ~
            exit_numbers:
              - "10"
          secondary_content: ~
          sub_content:
            text: ""
            maneuver_type: ~
            maneuver_modifier: ~
            roundabout_exit_degrees: ~
            lane_info:
              - active: false
                directions:
                  - straight
                active_direction: ~
              - active: false
                directions:
                  - straight
                active_direction: ~
              - active: false
                directions:
                  - straight
                  - slight right
                active_direction: slight right
              - active: true
                directions:
                  - slight right
                active_direction: slight right
            exit_numbers: []
          trigger_distance_before_maneuver: 264.647
      spoken_instructions: []
      annotations: redacted annotations json strings vec
      incidents: []
    - geometry:
        - lat: 45.375411
          lng: -122.586549
        - lat: 45.374846
          lng: -122.587677
        - lat: 45.374808
          lng: -122.587796
        - lat: 45.374781
          lng: -122.587928
        - lat: 45.374774
          lng: -122.588028
        - lat: 45.37478
          lng: -122.588136
        - lat: 45.374789
          lng: -122.588209
        - lat: 45.374812
          lng: -122.588294
        - lat: 45.374849
          lng: -122.588384
        - lat: 45.37488
          lng: -122.588451
        - lat: 45.374919
          lng: -122.588512
        - lat: 45.374978
          lng: -122.588569
        - lat: 45.375024
          lng: -122.58861
        - lat: 45.375075
          lng: -122.58864
        - lat: 45.375143
          lng: -122.588661
        - lat: 45.375324
          lng: -122.588722
        - lat: 45.375415
          lng: -122.588676
        - lat: 45.375483
          lng: -122.588617
        - lat: 45.375559
          lng: -122.58852
        - lat: 45.37561
          lng: -122.588432
        - lat: 45.375657
          lng: -122.58831
        - lat: 45.375703
          lng: -122.588114
      distance: 295
      duration: 12.258
      road_name: ""
      exits:
        - "10"
      instruction: Take exit 10 onto OR 213 South toward Oregon City/Molalla.
      visual_instructions:
        - primary_content:
            text: Cascade Highway
            maneuver_type: fork
            maneuver_modifier: slight right
            roundabout_exit_degrees: ~
            lane_info: ~
            exit_numbers: []
          secondary_content: ~
          sub_content:
            text: ""
            maneuver_type: ~
            maneuver_modifier: ~
            roundabout_exit_degrees: ~
            lane_info:
              - active: true
                directions:
                  - straight
                  - slight right
                active_direction: slight right
              - active: true
                directions:
                  - slight right
                active_direction: slight right
            exit_numbers: []
          trigger_distance_before_maneuver: 295
      spoken_instructions: []
      annotations: redacted annotations json strings vec
      incidents: []
    - geometry:
        - lat: 45.375703
          lng: -122.588114
        - lat: 45.375631
          lng: -122.58789
        - lat: 45.375576
          lng: -122.587713
        - lat: 45.375506
          lng: -122.58758
        - lat: 45.375468
          lng: -122.587524
        - lat: 45.37531
          lng: -122.587349
        - lat: 45.375003
          lng: -122.587226
        - lat: 45.374495
          lng: -122.587017
        - lat: 45.374137
          lng: -122.586867
        - lat: 45.373646
          lng: -122.586686
        - lat: 45.373572
          lng: -122.586658
        - lat: 45.373194
          lng: -122.586506
        - lat: 45.372855
          lng: -122.586367
        - lat: 45.372463
          lng: -122.586234
        - lat: 45.372156
          lng: -122.586125
        - lat: 45.371846
          lng: -122.586016
        - lat: 45.371508
          lng: -122.585901
      distance: 513.266
      duration: 25.269
      road_name: Cascade Highway
      exits: []
      instruction: Keep right to take OR 213 South/Cascade Highway.
      visual_instructions:
        - primary_content:
            text: You have arrived at your destination.
            maneuver_type: arrive
            maneuver_modifier: ~
            roundabout_exit_degrees: ~
            lane_info: ~
            exit_numbers: []
          secondary_content: ~
          sub_content: ~
          trigger_distance_before_maneuver: 513.266
      spoken_instructions: []
      annotations: redacted annotations json strings vec
      incidents: []
    - geometry:
        - lat: 45.371508
          lng: -122.585901
        - lat: 45.371508
          lng: -122.585901
      distance: 0
      duration: 0
      road_name: Cascade Highway
      exits: []
      instruction: You have arrived at your destination.
      visual_instructions: []
      spoken_instructions: []
      annotations: redacted annotations json strings vec
      incidents: []
